sympy/ntheory/qs.py

```diff
@@ -0,0 +1,391 @@
+        #if prime is 2 then sieve only with soln_1_p
+        for idx in range((M + factor.soln2) % factor.prime, 2*M, factor.prime):
+            sieve_array[idx] += factor.log_p
+    return sieve_array
+
+def check_smoothness(a, factor_base):
+    """Here we check that if `a` is a smooth number or not. If `a` is a smooth
+    number then it returns a vector of prime exponents modulo 2. For example
+    if a = 2 * 5**2 * 7**3 and the factor base contains {2, 3, 5, 7} then
+    `a` is a smooth number and this function returns ([1, 0, 0, 1], False). If
+    `a` is a partial relation which means that `a` a has one prime factor
+    greater than the factor_base then it returns `(a, True)` which denotes `a`
+    is a partial relation.
+
+    Parameters:
+    ===========
+
+    a : integr whose smootheness is to be checked
+    factor_base : factor_base primes
+    """
+    vec = []
+    if a < 0:
+        vec.append(1)
+        a *= -1
+    else:
+        vec.append(0)
+     #-1is not included in factor_base add -1 in vector
+    for factor in factor_base:
+        if a % factor.prime != 0:
+            vec.append(0)
+            continue
+        exp = 0
+        while a % factor.prime == 0:
+            exp += 1
+            a //= factor.prime
+        vec.append(exp % 2)
+    if a == 1:
+        return vec, False
+    if isprime(a):
+        return a, True
+    return None, None
+
+#Working fine
+def trial_division_stage(N, M, factor_base, sieve_array, sieve_poly, partial, ERROR_TERM):
+    """Compute trial division stage"""
+    sqrt_n = sqrt(float(N))
+    accumulated_val = log(M * sqrt_n)*2**10 - ERROR_TERM
+    smooth_relations = []
+    p_factor = set()
+    for idx, val in enumerate(sieve_array):
+        if val < accumulated_val:
+            continue
+        x = idx - M
+        v = sieve_poly.eval(x)
+        vec, ispartial = check_smoothness(v, factor_base)
+        if vec is None:
+            continue
+        u = sieve_poly.a*x + sieve_poly.b
+        if ispartial:
+            large_prime = vec
+            if large_prime not in partial:
+                partial[large_prime] = (u, v)
+                continue
+            else:
+                u_prev, v_prev = partial[large_prime]
+                partial.pop(large_prime)
+                try:
+                    large_prime_inv = mod_inverse(large_prime, N)
+                except ValueError:
+                    p_factor.add(large_prime)
+                    continue
+                u = u*u_prev*large_prime_inv
+                v = v*v_prev // (large_prime*large_prime)
+                vec, ispartial = check_smoothness(v, factor_base)
+        assert u*u % N == v % N
+        smooth_relations.append((u, v, vec))
+    return smooth_relations, p_factor
+
+
+#LINEAR ALGEBRA STAGE
+#Working fine
+def build_matrix(smooth_relations, factor_base):
+    matrix = []
+    for s_relation in smooth_relations:
+        matrix.append(s_relation[2])
+    return matrix
+
+#Working fine
```

Comment:
before we merge the take it as a reminder to remove these comments "working fine".